### PR TITLE
Port to syn 0.15

### PR DESCRIPTION
**Pull request by ogoffart**
Porting to newer syn is required in order to support new rust connstruct.

Fixes issue https://github.com/mystor/rust-cpp/issues/41

syn and proc_macro2 do not allow to get the line number or the verbatim
captrued text:  https://github.com/alexcrichton/proc-macro2/issues/110
This therefore has to redo the lexing manually to extract the relevant pieces
for which we need these information. Only the more comples parsing is still
done by syn.

This is a major change, so i took the opportunity to also run rustfmt

### CODE REVIEW

**Comment by ogoffart** [comment]
The CI is seems to be failling with nightly. But I cannot reproduce the failure locally when i install nightly.
It looks like for some reason the hash is not matching between cpp_build and cpp_macro for one function. I have no clue why that can be.

**Comment by ogoffart** [comment]
@mystor   any comments?

**Comment by mystor** [comment]
Hey, sorry I haven't had much time to look at this patch yet, especially with how large it is. Ill try to look at it this weekend but that's probably the earliest I can get to it.

**Comment by mystor** [comment]
I should also mention that syn 0.15 should be releasing soon (https://github.com/dtolnay/syn/issues/476), and it might be worthwhile to directly port to it. The entire parser engine has been replaced.

**Comment by ogoffart** [comment]
This means that cpp_common::parsing will need to be rewritten (again! :-))

However, I guess most of the rest of the code should not be affected, in particular the manual lexing will have to stay since proc_macro2 still does not allow to get the line information or offset.

I'll have a look at syn 0.15

**Comment by ogoffart** [comment]
I made the port to syn 0.15:  https://github.com/ogoffart/rust-cpp/pull/1  (It is based on this work)

File: cpp_build/src/lib.rs [hunk 1]
```diff
@@ -660,35 +617,16 @@ impl Config {
...
-            Err(err) => {
-                let mut err_s = err.to_string();
-                if let Some(i) = err_s.find("unparsed tokens after") {
-                    // Strip the long error message from syn
-                    err_s = err_s[0..i].to_owned();
-                }
-                warnln!(
-                    r#"-- rust-cpp parse error --
+        let mut visitor = parser::Parser::default();
+        if let Err(x) = visitor.parse_crate(&crate_root) {
```
Comment (mystor) [inline cpp_build/src/lib.rs]: nit: Use `err` here instead of `x`

File: cpp_build/src/lib.rs [hunk 2]
```diff
@@ -702,7 +640,6 @@ In order to provide a better error message, the build script will exit successfu
         if !self.std_flag_set {
             self.cc.flag_if_supported("-std=c++11");
         }
-
```
Comment (mystor) [inline cpp_build/src/lib.rs]: nit: unnecessary whitespace change

File: cpp_build/src/parser.rs [hunk 3]
```diff
@@ -0,0 +1,558 @@
+use cpp_common::{Class, Closure, Macro, RustInvocation};
```
Comment (mystor) [inline cpp_build/src/parser.rs]: I think we're going to have to do some level of custom parsing to keep working without a cpp fork, but I think we might be able to do it in a cleaner way. I'm probably going to take a bit to think about this.

File: cpp_common/src/lib.rs [hunk 4]
```diff
@@ -89,23 +101,67 @@ impl ClosureSig {
...
     pub sig: ClosureSig,
-    pub body: Spanned<String>,
+    pub body: TokenTree,
+    pub body_str: String, // with rust! macro replaced
     pub callback_offset: u32,
 }
 
+impl Parse for Closure {
+    fn parse(input: ParseStream) -> Result<Self> {
+        input.parse::<Option<Token![unsafe]>>()?;
```
Comment (mystor) [inline cpp_common/src/lib.rs]: Can we have more comments in the parsing code explaining what each block is trying to parse from the syntax :-)?

File: cpp_common/src/lib.rs [hunk 5]
```diff
@@ -89,23 +101,67 @@ impl ClosureSig {
...
+        } else {
+            (None, "void".to_owned())
+        };
+        let body = input.parse::<TokenTree>()?;
+        // Need to filter the spaces because there is a difference between
+        // proc_macro2 and proc_macro and the hashes would not match
+        let std_body = body
+            .to_string()
+            .chars()
+            .filter(|x| *x != ' ' && *x != '\n')
```
Comment (mystor) [inline cpp_common/src/lib.rs]: We could probably use 'is_whitespace' here. It might also be nice to do this filtering inside of `ClosureSig` instead, and have a single entry point.

File: cpp_common/src/lib.rs [hunk 6]
```diff
@@ -114,243 +170,115 @@ impl Class {
...
-            mac_body!($i, call!($e))
+impl Parse for RustInvocation {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let rust_token: Ident = input.parse()?;
+        if rust_token != "rust" {
+            return Err(syn::parse::Error::new(rust_token.span(), "expected `rust`"));
+        }
+        input.parse::<Token![!]>()?;
+        let content;
+        let p = parenthesized!(content in input);
```
Comment (mystor) [inline cpp_common/src/lib.rs]: nit: better variable names :-)

File: cpp_common/src/lib.rs [hunk 7]
```diff
@@ -114,243 +170,115 @@ impl Class {
...
+            return_type: if content.peek(Token![->]) {
+                content.parse::<Token![->]>()?;
+                content.parse::<Type>()?;
+                content.parse::<Token![as]>()?;
+                Some(content.parse::<syn::LitStr>()?.value())
+            } else {
+                None
+            },
         };
+        let c3;
```
Comment (mystor) [inline cpp_common/src/lib.rs]: Try doing `content.parse::<TokenTree>()?`

File: cpp_common/src/lib.rs [hunk 8]
```diff
@@ -114,243 +170,115 @@ impl Class {
...
-                alt!(punct!("[") | punct!("(") | punct!("{")),
-                $submac!($($args)*),
-                alt!(punct!("]") | punct!(")") | punct!("}"))
-            )
-        };
-        ($i:expr, $e:expr) => {
-            mac_body!($i, call!($e))
+impl Parse for RustInvocation {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let rust_token: Ident = input.parse()?;
```
Comment (mystor) [inline cpp_common/src/lib.rs]: Try using the new `custom_keyword!` here to make this cleaner :-)

File: cpp_macros/src/lib.rs [hunk 9]
```diff
@@ -193,8 +223,8 @@ NOTE: They cannot be generated by macro expansion."#,
     let mut call_args = Vec::new();
     for (i, capture) in closure.sig.captures.iter().enumerate() {
         let written_name = &capture.name;
-        let mac_name: Ident = format!("$var_{}", written_name).into();
-        let mac_cty: Ident = format!("$cty_{}", written_name).into();
+        let mac_name = Ident::new(&format!("var_{}", written_name), Span::call_site());
```
Comment (mystor) [inline cpp_macros/src/lib.rs]: It'd be nice if we could save the real span of our input written_name identifiers.

File: cpp_macros/src/lib.rs [hunk 10]
```diff
@@ -176,7 +205,8 @@ pub fn expand_internal(input: TokenStream) -> TokenStream {
     let source = input.to_string();
```
Comment (mystor) [inline cpp_macros/src/lib.rs]: One of the awesome things about the new `proc_macro2` is that we don't have to do this to-string dance anymore :-) - Just call `syn::parse<cpp_common::Closure>(input)`. Syn even provides an awesome `parse_macro_input` macro which even does error reporting for us! https://docs.rs/syn/0.15.3/syn/macro.parse_macro_input.html

File: cpp_macros/src/lib.rs [hunk 11]
```diff
@@ -279,10 +309,15 @@ NOTE: They cannot be generated by macro expansion."#,
         }
     };
 
-    let rust_invocations = parsing::find_all_rust_macro(synom::ParseState::new(&closure.body.node))
-        .expect("rust! macro");
+    use std::iter::FromIterator;
+    use syn::parse::Parser;
```
Comment (mystor) [inline cpp_macros/src/lib.rs]: Can you put these `use` statements at the top of the file?

File: cpp_macros/src/lib.rs [hunk 12]
```diff
@@ -279,10 +309,15 @@ NOTE: They cannot be generated by macro expansion."#,
...
+    let input = proc_macro2::TokenStream::from_iter(vec![closure.body].into_iter());
```
Comment (mystor) [inline cpp_macros/src/lib.rs]: nit: It's unfortunate you have to do `from_iter` here, but you should be able to just use a fixed size array rather than `vec!`

File: cpp_macros/src/lib.rs [hunk 13]
```diff
@@ -336,16 +371,18 @@ NOTE: They cannot be generated by macro expansion."#,
...
     #[cfg_attr(rustfmt, rustfmt_skip)]
     const SUFFIX: &'static [&'static str] = &[
         ")", ",", "0", ")", ".", "1", "]", ",", "}"
     ];
 
-    let s = source.find("stringify!(").expect("expected 'strignify!' token in class content") + 11;
-    let mut tokens : &str = &source[s..].trim();
+    let s = source
+        .find("stringify!(")
+        .expect("expected 'strignify!' token in class content") + 11;
```
Comment (mystor) [inline cpp_macros/src/lib.rs]: nit: stringify!

File: cpp_macros/src/lib.rs [hunk 14]
```diff
@@ -357,8 +394,8 @@ pub fn expand_wrap_class(input: TokenStream) -> TokenStream {
         tokens = &tokens[..tokens.len() - token.len()].trim();
     }
 
-    let class = parsing::cpp_class(synom::ParseState::new(tokens))
-        .expect("cpp_class! macro");
+    let input: ::proc_macro2::TokenStream = tokens.parse().unwrap();
+    let class = ::syn::parse2::<cpp_common::Class>(input).expect("cpp_class! macro");
```
Comment (mystor) [inline cpp_macros/src/lib.rs]: nit: It'd be nice if we could avoid the need to use so many prefixed `::` here.

File: cpp_common/src/lib.rs (hunk 8 above)
Comment (ogoffart) [reply cpp_common/src/lib.rs]: Right,  that did not exist when i did the port :-).

**Comment by ogoffart** [comment]
About the extraction of stringify! from cpp_macro, i implemented both alternative using explicit de-structuring (what is currently in this pull request) or use syn::visit to find the macro: https://github.com/ogoffart/rust-cpp/commit/0a75608e3e71c16bb7149de813454901fc12b778
I'm not sure what is the best. Using visit is less code and is cleaner. Probably resist better to changes in the cpp! and cpp_class! implementation, or in the syn AST. However it is probably less efficient, as it needs to re-visit the whole tree. And maybe it'd be a problem if one can have other macro (`cpp_class!(#[foo(stringify!(bar)) struct F as "F")` is not currently allowed, but maybe in the future?)

Overall, i think it is pretty important to upgrade `syn` as currently, using any of the new language constructs in a crate that wants to use cpp_build is impossible. That means already things like `impl Trait`or `dyn Trait`, use of the recently unreserved keywords, and probably more.
The situation is quite bad, as we only get a warning from cpp_build that there was a parse error, without telling you where or what, and the rustc compilation works only that the cpp! function are not found. or just link error if it was already compiled before.

For this reason, I think it is quite important to release a new version fast with an updated syn.

Regarding the parsing done in cpp_build, I think we should not wait for proc_macro2 to have what we need (verbatim source code, and line/column info), as this can take a long time. Forking is not a good idea as we would have the same problem again. 
The custom parsing code done in cpp_build is quite simple and limited to lexing. Originally, I thought to simply find occurence of the regexpt `\bcpp\s*!` within the source code, and then finding matching parentheses to get the end.  But that last part is impossible without proper lexing, as it can be quite complicated (example: `cpp!( ... { /* } */ " \" ) /*"; })`).
So I imported the lexing code from proc_macro2.  Change in the lexing rules might change un future rust versions, but are unlikely to change as much as the full grammar.

**Comment by mystor** [comment]
I'm fine with using the visitor, and it does seem quite a bit cleaner.

I think my worry with the need for manual tokenization is largely around the large complexity of having a complete copy of the lexing code within the crate. Perhaps we could make a separate crate which just handles string lexing and is used by rust-cpp, like the cpp_synmap crate used to be?

**Comment by ogoffart** [comment]
@mystor  What is the advantage to put it in a separated crate compared to a separated module?
Puting it in a different create means uploading another crate on crates.io, and also that means it needs somehow a public API and make it a full lexer.
Because right now, this is not even a full lexer. I do not do keywords, or numbers or lifetime or symbols, these are just ignored.

What exactly do you want in this crate?  Just put strnom.rs in its own crate?  Add `skip_literal` and `find_delimited`?

**Comment by mystor** [comment]
I think I'm okay with this grossness for now, but I'll think on how to clean it up more. Merge away and I can publish new versions

**Comment by ogoffart** [comment]
Thank you.

**Comment by mystor** [comment]
Published as `cpp 0.5` https://crates.io/crates/cpp. Thanks @ogoffart :-)
